File: number_baseball.cpp
#include<iostream>
#include<algorithm>
#include<string>
#include<vector>

using namespace std;

int solution(vector<vector<int>> baseball) {
    int answer = 0;
    return answer;
}

/*
   vector<vector<int>> chooseStrike(int sequence, int strike_number){
        // i개의 strike 숫자를 선택하여 가능한 숫자를 int vector로 반환한다. 
        // ex) strike_number : 2, sequence : 123 일 때, return : [[1,2],[1,3],[2,3]] 
        입력 수열을 문자 형태로 쪼갠다. ex) 123 -> '1','2','3'
        strike_number를 고려하여 순열을 생성한다. (순서 고려)
        생성한 순열을 이중 벡터에 저장하여 return 한다.

        return strike_arrays;
    }
*/

vector<vector<int>> combination(vector<int> list, int k){
    vector<vector<int>> result;

	// 0과1을 저장 할 벡터 생성
	vector<int> index;

	// k개의 1 추가
	for(int i=0; i<k; i++) index.push_back(1);
	
	// 2개(6개-2개)의 0 추가
	for(int i=0; i<list.size()-k; i++) index.push_back(0);

	//순열
	do{
        vector<int> tmp;
		for(int i=0; i<index.size(); i++)
			if(index[i] == 1)
				tmp.push_back(list[i]);

        result.push_back(tmp);
	}while(next_permutation(index.begin(), index.end()));
    return result;
}

vector<int> changeFormat(int sequence){
    // "123" -> [1,2,3]
    vector<int> tmp;
    for(int i = 0 ; i < 3 ;i++){
        int t = sequence%10;
        sequence /= 10;
        tmp.push_back(t);
    }
    return tmp;
}

vector<vector<int>> chooseStrike(int sequence, int strike_number){
    return combination(changeFormat(sequence), strike_number);
}

// array 에 있는 값들이 sequence에서 어떤 인덱스에 포함되어 있는지 vector<bool> 타입으로 반환한다.
vector<bool> informPosition(int sequence, vector<int> array){ 
    vector<int> sequence_v_type = changeFormat(sequence); // 123 -> [1,2,3]
    vector<bool> possible_check; // strike가 아닌 것은 false
    for(int i = 0 ; i < sequence_v_type.size(); i++) 
        possible_check.push_back(false);

    // sequence : [1,2,3] , strike_array : [2]가 주어질 때, 
    for(int i = 0 ; i< sequence_v_type.size(); i++)
        for(int j = 0 ; j<array.size(); j++)
            if(sequence_v_type[i] == array[j])
                possible_check[i] = true; // array 데이터가 sequence 몇 번째에 있는지를 나타낸다.

    return possible_check;
}
/*
    vector<vector<int>> chooseBall(int sequence, vector<int> strike_array, int ball_number){
        //입력으로 들어온 strike_array와 ball_number를 고려하여 가능한 ball의 조합을 vector 형식으로 표현한다.
        //ex) sequence : 123, strike_array : [1] , ball_number : 1 일때, return : [[2], [3]]
 
        return ball_arrays;
    }
*/
vector<vector<int>> chooseBall(int sequence, vector<int> strike_array, int ball_number){
    // sequence : [1,2,3], strike_array : [2], ball_number : 1 이면, [[1],[3]] 생성
    vector<int> mod_sequence_v_type;
    vector<int> sequence_v_type = changeFormat(sequence);
    vector<bool> ball_possible_check = informPosition(sequence, strike_array);

    for(int i = 0 ; i<sequence_v_type.size(); i++)
        if(ball_possible_check[i] == false)
            mod_sequence_v_type.push_back(sequence_v_type[i]);

    return combination(mod_sequence_v_type, ball_number);
}

typedef struct{
    int position;
    int value;
    char status;
}Info;

vector<Info> totalInformation(int sequence, vector<int> strike_array, vector<int> ball_array){
    /* 모든 정보를 다 담고 있는 벡터 리스트를 반환한다.
    result : [  [position, value, status], [position, value, status], ... ]
    */
    vector<int> sequence_v_type = changeFormat(sequence);
    vector<bool> strikeInfo = informPosition(sequence, strike_array);
    vector<bool> ballInfo = informPosition(sequence, ball_array);
    vector<Info> totalInfo;
    for(int i = 0 ; i < sequence_v_type.size(); i++){
        totalInfo[i].position = i;
        totalInfo[i].value = sequence_v_type[i];
        totalInfo[i].status = strikeInfo[i] ? 's' : ballInfo[i] ? 'b' : 'x';
    }
    return totalInfo;
}


vector<int> genSequences(int sequence, vector<int> strike_array, vector<int> ball_array){
    // total information을 가지고 가능한 모든 수열을 생성한다.
    // 여기서 작업 잠시 중지했습니다. 



}

int main(){
    vector<vector<int>> strike_arrays = chooseStrike(123, 2);
    return 0;
}

/*
    ## 알고리즘 구체화 ##



    vector<int> genSequences(vector<vector<int>> strike_arrays, vector<vector<int>> ball_arrays){
        //위 두 개의 함수에서 생성한 strike_arrays와 ball_arrays를 조합하여 가능한 수열을 생성한다.
    }

    void addSequence(vector<vector<int>> sequences){
        // 생성된 수열을 가능한 수열 리스트에 담아놓는다.
        // 담아 놓을 때, 기존에 있던 수열과 겹치지 않는 수열은 제거한다.
        // ex)
            input : 
            [
                [124,125]
                [143,153]
                [423,523]
            ]
            vector<int> global_vector :
            [124,165,177,143,153]

            updated global_vector :
            [124,143,153]
            }
        

    // 마지막으로, global_vector 안에 있는 원소의 갯수를 반환하면 된다.
    return global_vector.size();
    
*/

File: number_baseball.py
'''
    def chooseStrike(sequence, strike_number)
        output : possible strike lists

        ex)
        input : sequence = 123, strike_number = 1;
        output : [[1], [2], [3]]

    def chooseBall(sequence, strike_list, ball_number)
        sequence = 123, strike_list가 = [1]이 주어졌고, ball_number가 1이 주어진다면,
        이를 바탕으로 가능한 ball_lists를 출력한다.
        output : [[2],[3]]

    def genSequences(sequence, strike_list, ball_list) 
        ex) sequence = 123, strike_list = [1] , ball_list = [2] 와 같이 주어지면 
         1) strike_list에 있는 데이터는 고정시켜서 result_list에 저장. ex) result_list = [1, x, x]
         2) ball_list에 있는 데이터의 포지션 이외에 배치를 시킨다.
            즉, strike를 제외한 전체 포지션에서 가능한 순열 중,
            해당 위치에 데이터가 있는 것을 제거하면 된다. 
            ex) if sequence = 123, strike_list = [], ball_list = [1,2]
                일단 info_ball_list를 만든다. 이 list는 ball_list의 data가 sequence에서 어떤 index에 있는지를 담고 있다.
                    ex) info_ball_list = [[1,0], [2,1]]
                3자리 중, ball_number 만큼의 자리를 선택 후, 그 자리에 ball_list에 있는 데이터를 배치한다.
                모든 경우를 possible_sequences = [[1,2,x], [2,1,x], [1,x,2] , [2,x,1] , ... ] 와 같이 나타낸다.
                이 때, 각 possible_sequences를 info_ball_list와 비교해서 생성된 수열이 ball_list와 겹치는 것이 없는 경우에만 
                possible_sequence_forms에 추가한다.
                그러면 possible_sequence_forms = [[2,1,x], [2,x,1], ... ]와 같이 된다.
                다음으로 x 위치에 strike_list, ball_list에 없는 숫자들로 수열을 구성하여 result_list에 저장한다.
                result_list = [[214,215,216,217,218,219], [241,251,261,271,281,291], ...]
                return result_list

    def addSequences(result_list, final_result_list)
        if final_result_list == empty
            final_result_list = result_list
        else # 한 번 이상 경우의 수를 구한 경우
            final_result_list = final_result_list (교집합) result_list
        return final_result_list

    
    ## 전체 알고리즘
    def soultion(baseball):
        for i in range(0, len(baseball)):
            sequence = baseball[i][0]
            strike_number = baseball[i][1]
            ball_number = baseball[i][2]

            strike_list = chooseStrike(sequence, strike_number)
            ball_list = chooseBall(sequence, strike_list, ball_number)
            result_list = genSequences(sequence, strike_list, ball_list)
            final_result_list = addSequences(result_list, final_result_list)
            
        return len(final_result_list)
'''

from itertools import combinations

def chooseStrike(sequence, strike_number)
    '''
    output : possible strike lists

    ex)
    input : sequence = 123, strike_number = 1;
    output : [[1], [2], [3]]
    '''
    split_sequence = list(map(int, str(sequence))) # 123 -> [1,2,3]
    strike_lists = list(combinations(split_sequence,strike_number))
    return strike_lists

def chooseBall(sequence, strike_list, ball_number)
    '''
    sequence = 123, strike_list가 = [1]이 주어졌고, ball_number가 1이 주어진다면,
    이를 바탕으로 가능한 ball_lists를 출력한다.
    output : [[2],[3]]
    '''
    split_sequence = list(map(int, str(sequence))) #123 -> [1,2,3]
    remaining_number = list(set(split_sequence) - set(strike_list))
    ball_lists = list(combinations(split_sequence, ball_number))
    return ball_lists

 def genSequences(sequence, strike_list, ball_list) 
    '''
    ex) 
        sequence = 123, strike_list = [1] , ball_list = [2] 와 같이 주어지면 
        1) strike_list에 있는 데이터는 고정시켜서 result_list에 저장. ex) result_list = [1, x, x]
        2) ball_list에 있는 데이터의 포지션 이외에 배치를 시킨다.
        즉, strike를 제외한 전체 포지션에서 가능한 순열 중,
        해당 위치에 데이터가 있는 것을 제거하면 된다. 
        ex) if sequence = 123, strike_list = [], ball_list = [1,2]
            일단 info_ball_list를 만든다. 이 list는 ball_list의 data가 sequence에서 어떤 index에 있는지를 담고 있다.
                ex) info_ball_list = [[1,0], [2,1]]
            3자리 중, ball_number 만큼의 자리를 선택 후, 그 자리에 ball_list에 있는 데이터를 배치한다.
            모든 경우를 possible_sequences = [[1,2,x], [2,1,x], [1,x,2] , [2,x,1] , ... ] 와 같이 나타낸다.
            이 때, 각 possible_sequences를 info_ball_list와 비교해서 생성된 수열이 ball_list와 겹치는 것이 없는 경우에만 
            possible_sequence_forms에 추가한다.
            그러면 possible_sequence_forms = [[2,1,x], [2,x,1], ... ]와 같이 된다.
            다음으로 x 위치에 strike_list, ball_list에 없는 숫자들로 수열을 구성하여 result_list에 저장한다.
            result_list = [[214,215,216,217,218,219], [241,251,261,271,281,291], ...]
            return result_list
    '''
    split_sequence = list(map(int, str(sequence)))
    result_list = ['x','x','x']
    for strike_number in strike_list:
        for idx, number in enumerate(split_sequence):
            if strike_number == number:
                result_list[idx] = number
    
    
        



def solution(baseball):
    for i in range(0, len(baseball)):
        sequence = baseball[i][0]
        strike_number = baseball[i][1]
        ball_number = baseball[i][2]

        strike_lists = chooseStrike(sequence, strike_number)
        ball_lists = chooseBall(sequence, strike_list, ball_number)
        for strike_list in strike_lists:
            for ball_list in ball_lists:
                result_list = genSequences(sequence, strike_list, ball_list)
                final_result_list = addSequences(result_list, final_result_list)

    return len(final_result_list)



##test area
baseball = [[123,1,1], [356,1,0], [327,2,0], [489,0,1]]
# [sequence, strike_number, ball_number]
from itertools import combinations
list(map(int, str(12345)))
result = list(combinations([1,2,3,4],2))
print(result)

solution(baseball)




